gluster-exporter/metric_pv_lv_vg.go

```diff
@@ -0,0 +1,114 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"os/exec"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	log "github.com/sirupsen/logrus"
+)
+
+type collectDetails struct {
+	cmdStr string
+}
+
+// executes the command, and collect the details
+// returns the array of non empty output lines
+func (cd collectDetails) ExecCommand() ([]string, error) {
+	var outLines []string
+	cmdArr := strings.Fields(strings.TrimSpace(cd.cmdStr))
+	mainCmd, err := exec.LookPath(cmdArr[0])
+	if err != nil {
+		mainCmd = cmdArr[0]
+	}
+	outB, err := exec.Command(mainCmd, cmdArr[1:]...).Output()
+	if err != nil {
+		return nil, err
+	}
+	outR := bytes.NewReader(outB)
+	outScnr := bufio.NewScanner(outR)
+	// split each line
+	outScnr.Split(bufio.ScanLines)
+	log.Print("Output\n\n")
+	noOfLines := 0
+	for outScnr.Scan() {
+		line := strings.TrimSpace(outScnr.Text())
+		if line == "" {
+			continue
+		}
+		outLines = append(outLines, line)
+		noOfLines++
+	}
+	log.Println("No: of lines:", noOfLines)
+	return outLines, err
+}
+
+var (
+	metricLabels = []MetricLabel{
+		{
+			Name: "name",
+			Help: "Name of the metric, whose count is collected (PV or LV or VG etc)",
+		},
+	}
+
+	glusterPVCountMetric = newPrometheusGaugeVec(Metric{
+		Namespace: "gluster",
+		Name:      "pv_count",
+		Help:      "No: of physical volumes, got through pvs command",
+		LongHelp:  "",
+		Labels:    metricLabels,
+	})
+	glusterLVCountMetric = newPrometheusGaugeVec(Metric{
+		Namespace: "gluster",
+		Name:      "lv_count",
+		Help:      "No: of logical volumes, got through lvs command",
+		LongHelp:  "",
+		Labels:    metricLabels,
+	})
+	glusterVGCountMetric = newPrometheusGaugeVec(Metric{
+		Namespace: "gluster",
+		Name:      "vg_count",
+		Help:      "No: of volume groups, got through vgs command",
+		LongHelp:  "",
+		Labels:    metricLabels,
+	})
+)
+
+func pv_lv_vg_counts() (err error) {
+	collectDetailsOf := collectDetails{cmdStr: "pvs --noheading"}
```

Comment:
Instead of calling using 3 commands for pv, lv and vg count you can do it in one command `lvm vgs --noheading --reportformat=json -o vg_name, lv_count, pv_count` . U can declare a structure say:
```
type VGCount struct {
        VGName     `json:"vg_name"`
        LVCount    `json:"lv_count"`
        PVCount    `json:"pv_count"`
}
```
and parse the json output from the command into an array of VGCount structures say VGCounts just like in https://github.com/gluster/gluster-prometheus/blob/e714bb356000d27c5fa048dd8f6acdf364f3e050/gluster-exporter/metric_brick.go#L368-L382. Now to find the vg_count you can just find len(VGCounts).

Reply:
Thanks @sidharthanup , have modified the code accordingly...